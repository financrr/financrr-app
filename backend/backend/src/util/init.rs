use std::fmt::Debug;
use std::process;

use tracing::error;

pub(crate) fn expect_or_exit<T, E: Debug>(result: Result<T, E>, message: &str) -> T {
    match result {
        Ok(value) => value,
        Err(err) => {
            error!("{}: {:?}", message, err);
            process::exit(1);
        }
    }
}
