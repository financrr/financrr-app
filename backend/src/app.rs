use crate::initializers::openapi::OpenApiInitializer;
use crate::initializers::services::ServicesInitializer;
use crate::utils::folder::{create_necessary_folders, STORAGE_FOLDER};
use crate::{controllers, models::_entities::users, tasks, workers::downloader::DownloadWorker};
use async_trait::async_trait;
use loco_rs::cache::Cache;
use loco_rs::storage::Storage;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::{BackgroundWorker, Queue},
    boot::{create_app, BootResult, StartMode},
    cache,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    storage,
    task::Tasks,
    Result,
};
use migration::Migrator;
use mimalloc::MiMalloc;
use sea_orm::DatabaseConnection;
use std::path::Path;

#[global_allocator]
static GLOBAL: MiMalloc = MiMalloc;

pub struct App;

#[async_trait]
impl Hooks for App {
    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA").or(option_env!("GITHUB_SHA")).unwrap_or("dev")
        )
    }

    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    async fn boot(mode: StartMode, environment: &Environment) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![Box::new(OpenApiInitializer), Box::new(ServicesInitializer)])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .prefix("/api/v1")
            .add_route(controllers::auth::routes())
            .add_route(controllers::session::routes())
            .add_route(controllers::openapi::routes())
    }

    async fn after_context(ctx: AppContext) -> Result<AppContext> {
        create_necessary_folders()?;

        Ok(AppContext {
            cache: Cache::new(cache::drivers::inmem::new()).into(),
            storage: Storage::single(storage::drivers::local::new_with_prefix(STORAGE_FOLDER)?).into(),
            ..ctx
        })
    }

    async fn connect_workers(ctx: &AppContext, queue: &Queue) -> Result<()> {
        queue.register(DownloadWorker::build(ctx)).await?;
        Ok(())
    }
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::seed::SeedData);
        // tasks-inject (do not remove)
    }

    async fn truncate(db: &DatabaseConnection) -> Result<()> {
        // TODO add all other tables
        truncate_table(db, users::Entity).await?;
        Ok(())
    }

    async fn seed(db: &DatabaseConnection, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(db, &base.join("users.yaml").display().to_string()).await?;
        Ok(())
    }
}
