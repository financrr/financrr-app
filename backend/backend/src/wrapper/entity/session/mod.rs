use std::ops::Add;

use actix_web::dev::Payload;
use actix_web::{FromRequest, HttpRequest};
use futures_util::future::{join_all, LocalBoxFuture};
use sea_orm::{EntityName, EntityTrait, Set};
use serde::{Deserialize, Serialize};
use time::Duration as TimeDuration;
use time::OffsetDateTime;
use tokio::spawn;
use tokio::time::{sleep, Duration};
use tracing::{error, info};
use utoipa::ToSchema;
use uuid::Uuid;

use entity::session;
use entity::utility::time::get_now;

use crate::api::error::api::ApiError;
use crate::api::pagination::PageSizeParam;
use crate::config::Config;
use crate::database::entity::{count, delete, find_all_paginated, find_one_or_error, insert, update};
use crate::database::redis::{del, get, set_ex, zadd};
use crate::permission_impl;
use crate::util::auth::extract_bearer_token;
use crate::wrapper::entity::user::dto::Credentials;
use crate::wrapper::entity::user::User;
use crate::wrapper::entity::{TableName, WrapperEntity};
use crate::wrapper::permission::{Permission, Permissions};
use crate::wrapper::types::phantom::Identifiable;
use crate::wrapper::util::handle_async_result_vec;

pub(crate) mod dto;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub(crate) struct Session {
    pub(crate) id: i32,
    pub(crate) token: String,
    pub(crate) name: String,
    pub(crate) description: Option<String>,
    pub(crate) platform: Option<String>,
    #[serde(with = "time::serde::rfc3339")]
    pub(crate) expires_at: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub(crate) created_at: OffsetDateTime,
    pub(crate) user: User,
}

impl Session {
    pub(crate) async fn new(user: User, credentials: Credentials) -> Result<Self, ApiError> {
        let session_token = Self::generate_session_key();

        if Self::reached_session_limit(user.id).await? {
            Self::delete_oldest_session(user.id).await?;
        }

        // insert into database
        let session = session::ActiveModel {
            id: Default::default(),
            token: Set(session_token.clone()),
            name: Set(credentials.name),
            description: Set(credentials.description),
            platform: Set(credentials.platform),
            user: Set(user.id),
            created_at: Set(get_now()),
        };
        let model = insert(session).await?;

        let session = Self::from_model(model).await?;

        // insert into redis
        session.insert_into_redis().await?;

        //grant permissions to user
        session.add_permission(user.id, Permissions::all()).await?;

        Ok(session)
    }

    async fn insert_into_redis(&self) -> Result<(), ApiError> {
        set_ex(self.token.to_owned(), self.user.id.to_string(), self.expires_at.unix_timestamp() as u64).await?;
        zadd("sessions".to_owned(), self.token.to_owned(), self.expires_at.unix_timestamp() as f64).await?;

        Ok(())
    }

    pub(crate) async fn renew(mut self) -> Result<Self, ApiError> {
        self.expires_at = get_now().add(TimeDuration::hours(Config::get_config().session.lifetime_hours as i64));
        self.insert_into_redis().await?;

        let active_model = session::ActiveModel {
            id: Set(self.id),
            token: Set(self.token.to_owned()),
            name: Set(self.name.clone()),
            description: Set(self.description.clone()),
            platform: Set(self.platform.clone()),
            user: Set(self.user.id),
            created_at: Set(get_now()),
        };
        let model = update(active_model).await?;

        Self::from_model(model).await
    }

    pub(crate) async fn find_user_id(token: String) -> Result<i32, ApiError> {
        let user_id = Self::find_user_id_from_redis(token.to_owned()).await?;

        match user_id {
            Some(id) => Ok(id),
            None => Err(ApiError::InvalidSession()),
        }
    }

    async fn find_user_id_from_redis(token: String) -> Result<Option<i32>, ApiError> {
        let user_id = get::<Option<i32>>(token.to_owned()).await?;

        match user_id {
            Some(id) => Ok(Some(id)),
            None => Err(ApiError::InvalidSession()),
        }
    }

    pub(crate) async fn delete(self) -> Result<(), ApiError> {
        if let Err(e) = del(self.token.to_owned()).await {
            error!("Could not delete session {}: {}", self.token, e);
        }
        delete(session::Entity::delete_by_id(self.id)).await?;

        Ok(())
    }

    pub(crate) async fn delete_by_token(session_token: &String) -> Result<(), ApiError> {
        if let Err(e) = del(session_token.clone()).await {
            error!("Could not delete session {}: {}", session_token, e);
        }
        delete(session::Entity::delete_by_token(session_token.clone())).await?;

        Ok(())
    }

    pub(crate) async fn delete_all_with_user(user_id: i32) -> Result<(), ApiError> {
        let sessions = Self::find_all_by_user(user_id).await?;

        for session in sessions {
            session.delete().await?;
        }

        Ok(())
    }

    pub(crate) async fn find_all_paginated(page_size: &PageSizeParam) -> Result<Vec<Self>, ApiError> {
        let results =
            join_all(find_all_paginated(session::Entity::find(), page_size).await?.into_iter().map(Self::from_model))
                .await;

        handle_async_result_vec(results)
    }

    pub(crate) async fn find_all_by_user_paginated(
        user_id: i32,
        page_size: &PageSizeParam,
    ) -> Result<Vec<Self>, ApiError> {
        let results = join_all(
            find_all_paginated(session::Entity::find_by_user_id(user_id), page_size)
                .await?
                .into_iter()
                .map(Self::from_model),
        )
        .await;

        handle_async_result_vec(results)
    }

    pub(crate) async fn find_all_by_user(user_id: i32) -> Result<Vec<Self>, ApiError> {
        let results = join_all(
            find_all_paginated(session::Entity::find_by_user_id(user_id), &PageSizeParam::default())
                .await?
                .into_iter()
                .map(Self::from_model),
        )
        .await;

        handle_async_result_vec(results)
    }

    pub(crate) async fn find_by_id(id: i32) -> Result<Self, ApiError> {
        let model = find_one_or_error(session::Entity::find_by_id(id), "Session").await?;

        Self::from_model(model).await
    }

    pub(crate) async fn find_by_token(token: String) -> Result<Self, ApiError> {
        let model = find_one_or_error(session::Entity::find_by_token(token), "Session").await?;

        Self::from_model(model).await
    }

    pub(crate) async fn count_all() -> Result<u64, ApiError> {
        count(session::Entity::find()).await
    }

    pub(crate) async fn count_all_by_user(user_id: i32) -> Result<u64, ApiError> {
        count(session::Entity::find_by_user_id(user_id)).await
    }

    pub(crate) async fn reached_session_limit(user_id: i32) -> Result<bool, ApiError> {
        let sessions = Self::count_all_by_user(user_id).await?;

        Ok(sessions >= Config::get_config().session.limit)
    }

    pub(crate) async fn init() -> Result<(), ApiError> {
        // we have this set intentionally high because we have small datasets => we can afford to do this
        let limit: u64 = 1_000;
        let count = Self::count_all().await?;
        let pages = (count as f64 / limit as f64).ceil() as u64;

        info!("Loading {} pages with {} sessions per page...", pages, limit);
        for page in 1..=pages {
            info!("Loading page {}...", page);
            let page_size = PageSizeParam::new(page, limit);
            let sessions = Self::find_all_paginated(&page_size).await?;
            for session in sessions {
                session.insert_into_redis().await?;
                session.schedule_deletion().await?;
            }
        }

        Ok(())
    }

    async fn schedule_deletion(self) -> Result<(), ApiError> {
        let now = get_now().unix_timestamp();
        let expiration_timestamp = self.expires_at.unix_timestamp();
        let delay = expiration_timestamp - now;
        if delay > 0 {
            Self::schedule_deletion_task(self.token.to_owned(), delay as u64);
        } else {
            self.delete().await?;
        }

        Ok(())
    }

    fn schedule_deletion_task(session_key: String, delay: u64) {
        spawn(async move {
            sleep(Duration::from_secs(delay)).await;
            if let Err(e) = Self::delete_by_token(&session_key).await {
                error!("Could not delete session {}: {}", session_key, e);
            }
        });
    }

    async fn delete_oldest_session(user_id: i32) -> Result<(), ApiError> {
        let model = find_one_or_error(session::Entity::find_oldest_session_from_user_id(user_id), "Session").await?;
        let session = Self::from_model(model).await?;
        session.delete().await?;

        Ok(())
    }

    async fn from_model(model: session::Model) -> Result<Self, ApiError> {
        let user = User::find_by_id(model.user).await?;
        Ok(Self {
            id: model.id,
            name: model.name,
            description: model.description,
            platform: model.platform,
            token: model.token,
            user,
            expires_at: model.created_at.add(TimeDuration::hours(Config::get_config().session.lifetime_hours as i64)),
            created_at: model.created_at,
        })
    }

    fn generate_session_key() -> String {
        Uuid::new_v4().to_string()
    }
}

permission_impl!(Session);

impl TableName for Session {
    fn table_name() -> &'static str {
        session::Entity.table_name()
    }
}

impl WrapperEntity for Session {
    fn get_id(&self) -> i32 {
        self.id
    }
}

impl FromRequest for Session {
    type Error = ApiError;
    type Future = LocalBoxFuture<'static, Result<Self, ApiError>>;

    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
        let req = req.clone();
        Box::pin(async move {
            let token = extract_bearer_token(&req)?;

            Self::find_by_token(token).await
        })
    }
}
