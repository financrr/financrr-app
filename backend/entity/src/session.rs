//! `SeaORM` Entity, @generated by sea-orm-codegen 1.0.0-rc.4

use sea_orm::entity::prelude::*;
use sea_orm::{DeleteMany, QueryOrder, QuerySelect};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "session")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    #[sea_orm(column_type = "Text", unique)]
    pub token: String,
    #[sea_orm(column_type = "Text")]
    pub name: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub platform: Option<String>,
    pub user: i32,
    pub created_at: TimeDateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::user::Entity",
        from = "Column::User",
        to = "super::user::Column::Id",
        on_update = "Cascade",
        on_delete = "Cascade"
    )]
    User,
}

impl Related<super::user::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::User.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}

impl Entity {
    pub fn find_by_token(token: String) -> Select<Self> {
        Self::find().filter(Column::Token.eq(token))
    }

    pub fn find_by_user_id(user_id: i32) -> Select<Self> {
        Self::find().filter(Column::User.eq(user_id))
    }

    pub fn find_oldest_session_from_user_id(user_id: i32) -> Select<Self> {
        Self::find().filter(Column::User.eq(user_id)).order_by_asc(Column::CreatedAt).limit(1)
    }

    pub fn delete_by_token(session_token: String) -> DeleteMany<Self> {
        Self::delete_many().filter(Column::Token.contains(session_token))
    }
}
