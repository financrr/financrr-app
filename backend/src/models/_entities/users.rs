//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.3

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "users")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i64,
    #[sea_orm(column_type = "Text", unique)]
    pub email: String,
    #[sea_orm(column_type = "Text")]
    pub name: String,
    pub flags: i32,
    #[sea_orm(column_type = "Text")]
    pub password: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub reset_token: Option<String>,
    pub reset_sent_at: Option<DateTimeWithTimeZone>,
    #[sea_orm(column_type = "Text", nullable)]
    pub email_verification_token: Option<String>,
    pub email_verification_sent_at: Option<DateTimeWithTimeZone>,
    pub email_verified_at: Option<DateTimeWithTimeZone>,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::categories::Entity")]
    Categories,
    #[sea_orm(has_many = "super::currencies::Entity")]
    Currencies,
    #[sea_orm(has_many = "super::sessions::Entity")]
    Sessions,
    #[sea_orm(has_many = "super::tags::Entity")]
    Tags,
    #[sea_orm(has_many = "super::user_permissions::Entity")]
    UserPermissions,
}

impl Related<super::categories::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Categories.def()
    }
}

impl Related<super::currencies::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Currencies.def()
    }
}

impl Related<super::sessions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Sessions.def()
    }
}

impl Related<super::tags::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Tags.def()
    }
}

impl Related<super::user_permissions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::UserPermissions.def()
    }
}
