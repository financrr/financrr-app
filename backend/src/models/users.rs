pub use super::_entities::users::{self, ActiveModel, Entity, Model};
use crate::controllers::user::RegisterParams;
use crate::error::app_error::{AppError, AppResult};
use crate::models::_entities::sessions;
use crate::services::secret_generator::SecretGenerator;
use crate::services::snowflake_generator::SnowflakeGenerator;
use chrono::offset::Local;
use enumflags2::_internal::RawBitFlags;
use enumflags2::bitflags;
use loco_rs::{hash, prelude::*};
use sea_orm::prelude::Expr;
use sea_orm::sea_query::IntoCondition;
use sea_orm::{JoinType, PaginatorTrait, QuerySelect, RelationTrait};

#[bitflags(default = User)]
#[repr(u8)]
#[derive(Copy, Clone, Debug, PartialEq)]
pub enum UserFlags {
    Admin = 0b0001,
    User = 0b0010,
}

#[async_trait::async_trait]
impl ActiveModelBehavior for super::_entities::users::ActiveModel {
    async fn before_save<C>(mut self, _db: &C, insert: bool) -> Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        if !self.updated_at.is_set() {
            self.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        if insert {
            self.created_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        Ok(self)
    }
}

impl super::_entities::users::Model {
    /// finds a user by the provided email
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_email(db: &DatabaseConnection, email: &str) -> AppResult<Option<Self>> {
        let user = users::Entity::find()
            .filter(query::condition().eq(users::Column::Email, email).build())
            .one(db)
            .await?;

        Ok(user)
    }

    pub async fn is_email_unique(db: &DatabaseConnection, email: &str) -> AppResult<bool> {
        let user = users::Entity::find()
            .select_only()
            .column(users::Column::Email)
            .filter(users::Column::Email.eq(email))
            .count(db)
            .await?;

        Ok(user == 0)
    }

    /// finds a user by the provided verification token
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_verification_token(
        db: &DatabaseConnection,
        email: &str,
        token: &str,
    ) -> Result<Option<Self>, AppError> {
        Ok(users::Entity::find()
            .filter(
                query::condition()
                    .eq(users::Column::Email, email)
                    .eq(users::Column::EmailVerificationToken, token)
                    .build(),
            )
            .one(db)
            .await?)
    }

    /// finds a user by the provided reset token
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_reset_token(db: &DatabaseConnection, email: &str, token: &str) -> ModelResult<Self> {
        let user = users::Entity::find()
            .filter(
                query::condition()
                    .eq(users::Column::Email, email)
                    .eq(users::Column::ResetToken, token)
                    .build(),
            )
            .one(db)
            .await?;
        user.ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_id(db: &DatabaseConnection, id: i64) -> ModelResult<Self> {
        let user = users::Entity::find()
            .filter(query::condition().eq(users::Column::Id, id).build())
            .one(db)
            .await?;
        user.ok_or_else(|| ModelError::EntityNotFound)
    }

    /// finds a user by the provided api key
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_api_key(db: &DatabaseConnection, api_key: &str) -> ModelResult<Self> {
        let cloned_api_key = api_key.to_string();
        let user = users::Entity::find()
            .join(
                JoinType::LeftJoin,
                sessions::Relation::Users.def().rev().on_condition(move |_left, right| {
                    Expr::col((right, sessions::Column::ApiKey))
                        .eq(cloned_api_key.as_str())
                        .into_condition()
                }),
            )
            .one(db)
            .await?;
        user.ok_or_else(|| ModelError::EntityNotFound)
    }

    /// Verifies whether the provided plain password matches the hashed password
    ///
    /// # Errors
    ///
    /// when could not verify password
    #[must_use]
    pub fn verify_password(&self, password: &str) -> bool {
        hash::verify_password(password, &self.password)
    }

    /// Asynchronously creates a user with a password and saves it to the
    /// database.
    ///
    /// # Errors
    ///
    /// When could not save the user into the DB
    pub async fn create_with_password(
        db: &DatabaseConnection,
        snowflake_generator: &SnowflakeGenerator,
        params: &RegisterParams,
    ) -> Result<Self, AppError> {
        let txn = db.begin().await?;

        let password_hash = hash::hash_password(&params.password)?;
        let user = users::ActiveModel {
            id: ActiveValue::set(snowflake_generator.next_id().map_err(|e| ModelError::Any(e.into()))?),
            email: ActiveValue::set(params.email.to_string()),
            password: ActiveValue::set(password_hash),
            reset_token: Default::default(),
            reset_sent_at: Default::default(),
            email_verification_token: Default::default(),
            email_verification_sent_at: Default::default(),
            email_verified_at: Default::default(),
            name: ActiveValue::set(params.name.to_string()),
            flags: ActiveValue::set(UserFlags::DEFAULT as i32),
            created_at: Default::default(),
            updated_at: Default::default(),
        }
        .insert(&txn)
        .await?;

        txn.commit().await?;

        Ok(user)
    }
}

impl super::_entities::users::ActiveModel {
    /// Sets the email verification information for the user and
    /// updates it in the database.
    ///
    /// This method is used to record the timestamp when the email verification
    /// was sent and generate a unique verification token for the user.
    ///
    /// # Errors
    ///
    /// when has DB query error
    pub async fn set_email_verification_sent(
        mut self,
        db: &DatabaseConnection,
        secret_generator: &SecretGenerator,
    ) -> ModelResult<Model> {
        const EMAIL_VERIFICATION_TOKEN_LENGTH: usize = 8;
        self.email_verification_sent_at = ActiveValue::set(Some(Local::now().into()));
        self.email_verification_token = ActiveValue::Set(Some(
            secret_generator.generate_token_with_length(EMAIL_VERIFICATION_TOKEN_LENGTH),
        ));
        Ok(self.update(db).await?)
    }

    /// Sets the information for a reset password request,
    /// generates a unique reset password token, and updates it in the
    /// database.
    ///
    /// This method records the timestamp when the reset password token is sent
    /// and generates a unique token for the user.
    ///
    /// # Arguments
    ///
    /// # Errors
    ///
    /// when has DB query error
    pub async fn set_forgot_password_sent(
        mut self,
        db: &DatabaseConnection,
        secret_generator: &SecretGenerator,
    ) -> ModelResult<Model> {
        const RESET_TOKEN_LENGTH: usize = 8;
        self.reset_sent_at = ActiveValue::set(Some(Local::now().into()));
        self.reset_token = ActiveValue::Set(Some(secret_generator.generate_token_with_length(RESET_TOKEN_LENGTH)));
        Ok(self.update(db).await?)
    }

    /// Records the verification time when a user verifies their
    /// email and updates it in the database.
    ///
    /// This method sets the timestamp when the user successfully verifies their
    /// email.
    ///
    /// # Errors
    ///
    /// when has DB query error
    pub async fn verified(mut self, db: &DatabaseConnection) -> ModelResult<Model> {
        self.email_verified_at = ActiveValue::set(Some(Local::now().into()));
        self.email_verification_token = ActiveValue::Set(None);
        Ok(self.update(db).await?)
    }

    /// Resets the current user password with a new password and
    /// updates it in the database.
    ///
    /// This method hashes the provided password and sets it as the new password
    /// for the user.
    ///
    /// # Errors
    ///
    /// when has DB query error or could not hashed the given password
    pub async fn reset_password(mut self, db: &DatabaseConnection, password: &str) -> ModelResult<Model> {
        self.password = ActiveValue::set(hash::hash_password(password).map_err(|e| ModelError::Any(e.into()))?);
        self.reset_token = ActiveValue::Set(None);
        self.reset_sent_at = ActiveValue::Set(None);
        Ok(self.update(db).await?)
    }
}
