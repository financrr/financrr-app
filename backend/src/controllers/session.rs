use crate::controllers::user::{MAX_PASSWORD_LENGTH, MIN_PASSWORD_LENGTH};
use crate::error::app_error::{
    AppError, AppResult, GeneralInternalServerErrorResponse, InvalidBearerTokenResponse, InvalidEmailOrPasswordResponse,
};
use crate::middlewares::authentication::Authenticated;
use crate::models::_entities::sessions;
use crate::models::users;
use crate::services::secret_generator::SecretGenerator;
use crate::services::snowflake_generator::SnowflakeGenerator;
use crate::views::session::SessionResponse;
use axum::extract::State;
use axum::http::StatusCode;
use axum::routing::{get, post};
use axum::{debug_handler, Extension, Json};
use loco_rs::app::AppContext;
use loco_rs::controller::Routes;
use serde::Deserialize;
use utoipa::ToSchema;
use validator::Validate;

#[derive(Debug, Deserialize, ToSchema, Validate)]
pub struct CreateSessionParams {
    #[validate(email)]
    pub email: String,
    #[validate(length(min = "MIN_PASSWORD_LENGTH", max = "MAX_PASSWORD_LENGTH"))]
    pub password: String,
    pub name: Option<String>,
    pub user_agent: Option<String>,
}

/// Login/Create a new Session.
///
/// This endpoint is used to create a new Session for a User.
/// Returns a new Session with a Bearer Token that can be used to authenticate the User.
#[utoipa::path(post,
    path = "/api/v1/sessions",
    tag = "Session",
    responses(
        (status = StatusCode::CREATED, description = "Successfully created a new Session.", content_type="application/json", body = SessionResponse),
        InvalidEmailOrPasswordResponse,
        GeneralInternalServerErrorResponse,
    ),
)]
#[debug_handler]
async fn create(
    State(ctx): State<AppContext>,
    Extension(snowflake_generator): Extension<SnowflakeGenerator>,
    Extension(secret_generator): Extension<SecretGenerator>,
    Json(params): Json<CreateSessionParams>,
) -> AppResult<(StatusCode, Json<SessionResponse>)> {
    params.validate()?;
    let user = users::Model::find_by_email(&ctx.db, &params.email).await?;

    let user = match user {
        None => return Err(AppError::InvalidEmailOrPassword())?,
        Some(user) => user,
    };

    if user.email_verified_at.is_none() {
        return Err(AppError::EmailNotVerified())?;
    }

    if !user.verify_password(&params.password) {
        return Err(AppError::InvalidEmailOrPassword())?;
    }

    let session =
        sessions::Model::create_session(&ctx.db, &snowflake_generator, &secret_generator, &user, &params).await?;

    Ok((StatusCode::CREATED, Json(SessionResponse::from((session, user)))))
}

/// Retrieve the current Session.
#[utoipa::path(get,
    path = "/api/v1/sessions/current",
    tag = "Session",
    responses(
        (status = StatusCode::OK, description = "Successfully retrieved the current Session.", content_type="application/json", body = SessionResponse),
        GeneralInternalServerErrorResponse,
        InvalidBearerTokenResponse,
    ),
    security(
        ("bearer_token" = [])
    ),
)]
#[debug_handler]
async fn current(
    State(ctx): State<AppContext>,
    Authenticated(session): Authenticated<sessions::Model>,
) -> AppResult<(StatusCode, Json<SessionResponse>)> {
    let user = users::Model::find_by_id(&ctx.db, session.user_id).await?;

    Ok((StatusCode::OK, Json(SessionResponse::from((session, user)))))
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("/sessions")
        .add("/", post(create))
        .add("/current", get(current))
}
