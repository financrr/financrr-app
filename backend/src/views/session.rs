use crate::models::_entities::sessions;
use crate::models::_entities::sessions::Model;
use crate::models::users;
use crate::types::snowflake::Snowflake;
use crate::views::user::UserResponse;
use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct SessionResponse {
    pub id: Snowflake,
    pub user: UserResponse,
    pub api_key: String,
    pub name: Option<String>,
    pub user_agent: Option<String>,
    pub last_accessed_at: Option<DateTime<FixedOffset>>,
    pub created_at: DateTime<FixedOffset>,
    pub updated_at: DateTime<FixedOffset>,
}

impl From<(sessions::Model, users::Model)> for SessionResponse {
    fn from(value: (Model, users::Model)) -> Self {
        let user = UserResponse::from(value.1);
        let session = value.0;
        Self {
            id: Snowflake::new(session.id),
            user,
            api_key: session.api_key,
            name: session.name,
            user_agent: session.user_agent,
            last_accessed_at: session.last_accessed_at,
            created_at: session.created_at,
            updated_at: session.updated_at,
        }
    }
}
