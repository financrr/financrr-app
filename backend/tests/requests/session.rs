use crate::helpers::init::init_test;
use crate::helpers::session::clean_up_session_response;
use crate::helpers::users::{create_unverified_user_with_email, create_user_with_email, DEFAULT_PASSWORD};
use axum::http::StatusCode;
use financrr::app::App;
use financrr::views::session::SessionResponse;
use insta::{assert_json_snapshot, with_settings};
use loco_rs::testing;
use rstest::rstest;
use sea_orm::IntoActiveModel;
use serde_json::json;
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("sessions_request");
        let _guard = settings.bind_to_scope();
    };
}

#[rstest]
#[case("can_login_with_valid_password", DEFAULT_PASSWORD, StatusCode::CREATED)]
#[case("can_login_with_invalid_password", "invalid-password", StatusCode::BAD_REQUEST)]
#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_login(#[case] test_name: &str, #[case] password: &str, #[case] expected_status_code: StatusCode) {
    init_test!();

    testing::request::<App, _, _>(|request, ctx| async move {
        const EMAIL: &str = "can.login@financrr.test";
        let _ = create_user_with_email(&ctx, EMAIL).await;

        let payload = json!({
            "email": EMAIL,
            "password": password,
        });
        let response = request.post("/api/v1/sessions").json(&payload).await;
        assert_eq!(response.status_code(), expected_status_code);

        if StatusCode::CREATED == expected_status_code {
            let session_response: SessionResponse = response.json();

            with_settings!({
                filters => clean_up_session_response()
            }, {
                assert_json_snapshot!(test_name, session_response);
            })
        }
    })
    .await;
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_login_without_verify() {
    init_test!();

    testing::request::<App, _, _>(|request, ctx| async move {
        const EMAIL: &str = "can.login.without.verify@financrr.test";
        let user = create_unverified_user_with_email(&ctx, EMAIL).await;

        let payload = json!({
            "email": EMAIL,
            "password": DEFAULT_PASSWORD,
        });
        let response = request.post("/api/v1/sessions").json(&payload).await;
        assert_eq!(response.status_code(), StatusCode::BAD_REQUEST);

        let _ = user.into_active_model().verified(&ctx.db).await.unwrap();
        let response = request.post("/api/v1/sessions").json(&payload).await;
        assert_eq!(response.status_code(), StatusCode::CREATED);

        let session_response: SessionResponse = response.json();
        with_settings!({
            filters => clean_up_session_response()
        }, {
            assert_json_snapshot!("can_login_without_verify", session_response);
        });
    })
    .await;
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_get_current_session() {
    init_test!();

    testing::request::<App, _, _>(|request, ctx| async move {
        const EMAIL: &str = "can.get.current.session@financrr.test";
        let user = create_user_with_email(&ctx, EMAIL).await;

        let payload = json!({
            "email": user.email,
            "password": DEFAULT_PASSWORD,
        });
        let response = request.post("/api/v1/sessions").json(&payload).await;
        assert_eq!(response.status_code(), StatusCode::CREATED);

        let session_response: SessionResponse = response.json();
        let response = request
            .get("/api/v1/sessions/current")
            .add_header("Authorization", format!("Bearer {}", session_response.api_key))
            .await;
        assert_eq!(response.status_code(), StatusCode::OK);

        let session_response: SessionResponse = response.json();
        with_settings!({
            filters => clean_up_session_response()
        }, {
            assert_json_snapshot!("can_get_current_user", session_response);
        });
    })
    .await;
}
