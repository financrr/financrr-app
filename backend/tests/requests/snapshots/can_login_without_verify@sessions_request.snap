---
source: tests/requests/session.rs
assertion_line: 79
expression: session_response
snapshot_kind: text
---
{
  "id": ID,
  "user": {
    "id": ID,
    "email": "can.login.without.verify@financrr.test",
    "name": "Test Account",
    "flags": 2,
    "email_verification_sent_at": "DATEZ",
    "email_verified_at": "DATEZ",
    "created_at": "DATEZ",
    "updated_at": "DATEZ"
  },
  "api_key": KEY,
  "name": null,
  "user_agent": null,
  "last_accessed_at": null,
  "created_at": "DATEZ",
  "updated_at": "DATEZ"
}
